### PR TITLE
Feature/separate ocl kernels

### PR DESCRIPTION
**Pull request by vbkaisetsu**
This branch separates OpenCL kernels into `opencl_device_kernel.cl`.

`opencl_device_kernel.cl` will be converted to `char[]` by a CMake script before compile `opencl_device.cc`.

### CODE REVIEW

File: cmake/generate_char_array.cmake [hunk 1]
```diff
@@ -0,0 +1,4 @@
+file(READ ${CMAKE_ARGV3} hexdata HEX)
```
Comment (odashi) [inline cmake/generate_char_array.cmake]: `hexdata` -> `hex_data` to adjust variable naming

Comment (odashi) [inline cmake/generate_char_array.cmake]: Add enough manual about description and usage for this script.

File: primitiv/CMakeLists.txt [hunk 2]
```diff
@@ -103,6 +103,16 @@ endif()
 
 # Build rules of the OpenCL backend.
 if(PRIMITIV_USE_OPENCL)
+  add_custom_command(OUTPUT opencl_device_kernel.h
```
Comment (odashi) [inline primitiv/CMakeLists.txt]: `opencl_device_kernel.h` -> `${CMAKE_CURRENT_BINARY_DIR}/opencl_device_kernels.inc`.
The extension `.h` should represent *header* files which describe APIs of the library.

File: primitiv/CMakeLists.txt [hunk 3]
```diff
@@ -103,6 +103,16 @@ endif()
...
+    COMMAND
+      ${CMAKE_COMMAND}
+      -P
+      ${PROJECT_SOURCE_DIR}/cmake/generate_char_array.cmake
+      ${PROJECT_SOURCE_DIR}/primitiv/opencl_device_kernel.cl
```
Comment (odashi) [inline primitiv/CMakeLists.txt]: `${PROJECT_SOURCE_DIR}` -> `${CMAKE_CURRENT_SOURCE_DIR}`

File: primitiv/opencl_device.cc [hunk 4]
```diff
@@ -98,535 +98,12 @@ std::uint32_t calc_num_blocks(std::uint32_t size, std::uint32_t num_threads) {
...
-  const unsigned i = get_global_id(0);
-  const unsigned bid_y = get_group_id(1);
-  const unsigned shift = bid_y * size;
-  if (i < size) atomic_add_float(py + i + mby * shift, -px[i + mbx * shift]);
-}
-)EOS";
-
-  return ss.str();
+const std::string generate_kernels() {
+  const std::string ss({
```
Comment (odashi) [inline primitiv/opencl_device.cc]: Not necessary to use the temporary variable `ss` and the resulting string can be directly returned.

File: primitiv/opencl_device.cc [hunk 5]
```diff
@@ -739,7 +216,14 @@ struct OpenCLInternalState {
           // Then, we can delete the buffer safely.
           delete static_cast<cl::Buffer *>(ptr);
         }) {
-      cl::Program program(context, ::generate_kernels(), true);
+      cl::Program program(context, ::generate_kernels());
+      try {
+        program.build({device});
+      } catch (...) {
+        std::cerr << program.getBuildInfo<CL_PROGRAM_BUILD_LOG>(device)
```
Comment (odashi) [inline primitiv/opencl_device.cc]: Should not directly print any messages to the standard streams without users' request.
Using `THROW_ERROR` instead to inform this message.

File: primitiv/opencl_device.cc [hunk 6]
```diff
@@ -98,535 +98,12 @@ std::uint32_t calc_num_blocks(std::uint32_t size, std::uint32_t num_threads) {
...
-    const unsigned mbx, const unsigned mby, global float *py) {
-  const unsigned i = get_global_id(0);
-  const unsigned bid_y = get_group_id(1);
-  const unsigned shift = bid_y * size;
-  if (i < size) atomic_add_float(py + i + mby * shift, -px[i + mbx * shift]);
-}
-)EOS";
-
-  return ss.str();
+const std::string generate_kernels() {
```
Comment (odashi) [inline primitiv/opencl_device.cc]: Remove const qualifier.

**Comment by vbkaisetsu** [comment]
Fixed.

**Pull request merged**